Tensor API Refactoring

karllessard opened the pull request.

alextp committed:
changes

alextp committed:
changes

alextp committed:
fix link

alextp committed:
new-rfc

Craigacp commented on sigs/jvm/rfcs/20190904-java-tensors.md:
```diff
@@ -0,0 +1,301 @@
+accessing a tensor instance. Actually, this information is carried by the [`DataType` enum class](https://github.com/karllessard/tensorflow/blob/master/tensorflow/java/src/main/java/org/tensorflow/DataType.java),
+which converts back and forth a type alias (such as `Integer`) to a TF data type (such as `INT32`).
+
+Again, this conversion can be sometimes painful and could be avoid by changing the `DataType` enum class to a 
```
I'm confused as to how this is better than an enum. The enum can carry all this information already, and is just as accessible from JNI as an anonymous class is if not more. Do you have an example which shows the benefit of this approach over the current one? The placeholder factory could be modified to ask the enum what the ordinal is (in the way it would need to be updated to call DataType.ordinal).

karllessard replied:
Ok, maybe I mixed two different concept here.

The main idea was not to get rid of the enum but to get rid of the conversion between a Java type and an enum, using [this method](https://github.com/tensorflow/tensorflow/blob/a2048b0a19157054a6bca25610c8b7aa14bef2d5/tensorflow/java/src/main/java/org/tensorflow/DataType.java#L96), like [here](https://github.com/tensorflow/tensorflow/blob/a2048b0a19157054a6bca25610c8b7aa14bef2d5/tensorflow/java/src/main/java/org/tensorflow/op/core/Zeros.java#L50).

Now why using a static class instead of an enum? Well, since we already have a specific class per supported tensor type (in this RFC I mean), for me it sounds just more intuitive to store all information related to that type inside this class and the user does not have to deduce the corresponding value in an enum. (e.g. the type info of `TInt32`is `TInt32.DTYPE`, instead of the type of of `TInt32`is `DataType.INT32`).

Craigacp replied:
Ok. That approach precludes switching over the types or other things which need to know about all the Tensor types in the C API. Which might be fine, as you point out later it's much more extensible, but I'm not sure if it's worth exposing types in Java that aren't understood by the C API underneath.

karllessard replied:
> but I'm not sure if it's worth exposing types in Java that aren't understood by the C API underneath.

That is not the plan neither, all tensor class such a `TInt32`would match a type supported by the C API. Or maybe you mean the "custom" type I was proposing later? Yes, this one need more rethinking for sure.

Craigacp replied:
Yeah, I mean the custom types proposed at the end of the document.

Craigacp replied:
However there are things like fp16 and bf16 which aren't easily representable in Java at the moment, but are probably useful.

Craigacp commented on sigs/jvm/rfcs/20190904-java-tensors.md:
```diff
@@ -0,0 +1,301 @@
+       return allocate(DTYPE, Shape.make()).set(value);
+   }
+   
+   public static TInt32 vector(Integer... values) {
```
The boxing here makes this very inefficient, and also it's not possible to pass in an `int[]`. We'd need a signature that is `int...` in addition to these methods.

karllessard replied:
Can you please give more explanation what makes the boxed varargs not performant, other than the regular required conversion? I'm lacking a bit of knowledge in this field and I think you know a lot more about it. 

I'm just trying to avoid supporting primitive arrays in the `NdArray` API, like I used to, because it complexifies a lot the hierarchy of objects required if we want to keep control on which tensor supports which primitive array (i.e. you can pass a `int[]` to a `TInt32` but not to a `TBool`).

Craigacp replied:
The varargs construction allocates an array and writes all the values into it. So for an `f(Integer... inputs)` called as `obj.f(1000,2000,3000,4000)` it first creates an Integer object which contains the field storing 1000, then an Integer representing 2000, etc. Then it stores a reference to each of those Integers into the created Integer[4]. This means that the call allocated 5 objects on the heap. For an entry point `g(int... inputs)` called as `obj.g(1000,2000,3000,4000)` then it allocates a single int[4] and writes all the values into that array. So that's only a single object allocation. This isn't too important for short arrays like in your example, but it's really important for longer arrays.

The integers used are important too, as the JDK has a cache of the integers in the range [-127, 127] so those integers shouldn't trigger allocation. But values outside those ranges (and Floats, Doubles etc) will allocate a fresh object.

The code to convert an array of primitives into an array of unboxed primitives is a single line `IntStream.of(intArr).boxed().toArray(Integer[]::new)` and there isn't a FloatStream or BooleanStream to do the equivalent transformation for those types quickly.

I think it's important to be able to pass in a primitive array, as that's what people will be using for efficiency reasons, and if we require the intermediate boxing then it's going to generate a lot of garbage and object allocations.

Either that or we get the NDArray set(T value, int... indices) method to be extremely fast and remove all ways of creating things with arrays or varargs, so the natural way of using it is generating a blank one and then filling it with the feature extractor. If it's backed by a ByteBuffer then it's probably a quick method, but if it's writing directly into the TF Tensor then we pay the JNI cost every single time.

The split between primitives and objects is painful in current Java, as you need to use the primitives to get any kind of speed, but the lack of generics over primitives leads to ugly APIs with edges. Once specialised generics lands we should be able to get around this, but that's still a while away.

karllessard replied:
Thanks a lot, that was very helpful. And yes, your last comment says it all:

> The split between primitives and objects is painful in current Java, as you need to use the primitives to get any kind of speed, but the lack of generics over primitives leads to ugly APIs with edges.

I guess we can start with an easy clean API principally relying on boxed types, while being careful to avoid allocation overhead like you stated, and then add the primitive arrays if performances are not reached. But like you also said: 

> This isn't too important for short arrays like in your example, but it's really important for longer arrays.

which is probably most of the cases when an `NdArray` using one of those method (`vector()`, `set()`, ...). For bulk initialize of a large set of data, probably the `NdArray.write(DataBuffer)`will be a better pick.

> If it's backed by a ByteBuffer then it's probably a quick method, but if it's writing directly into the TF Tensor then we pay the JNI cost every single time.

There should not be any difference in performances between writing to a `NdArray`or a `Tensor`as both are backed by `ByteBuffer` (currently JNI is only implied to return a reference to the `ByteBuffer`of the tensor and then everything can be done in Java space).

Craigacp replied:
When does the ByteBuffer get handed to the TF C API?

karllessard replied:
Actually? When the tensor is allocated by the native layer, its buffer is mapped to a `ByteBuffer` (see [here](https://github.com/tensorflow/tensorflow/blob/e1667ec8563262e7fe6cd07ddc13af599e00e906/tensorflow/java/src/main/native/tensor_jni.cc#L483) and returned to the Java runtime.

So it was just per design that a user can not write directly to the tensor memory right now, something we’ll change with the new Tensor NIO interface.

BTW, @eronwright, who wrote  that code a few years ago, has join the SIG lately to offer his help on this.

Craigacp replied:
Ah, cool, I didn't realize that. Does that interfere with any of the automatic memory management it's doing under the hood (e.g. CUDA unified memory, or whatever magic happens with TPUs)?

karllessard replied:
My blind guess is that a tensor in GPU memory is always copy to local memory before being mapped to a ByteBuffer, but Eron probably know better about it.

If that's the case, then yes there would be a performance impact caused by the copy so we should make sure to map that memory only when required (which is the case right now... I think).

EronWright replied:
I wouldn't be surprised if a direct `ByteBuffer` could manipulate a GPU's mapped memory.   Not an expert on that.

About the whole concept of an `NDArray` that is backed by a direct byte buffer, we might find inspiration in the `V8TypedArray` in the Java bindings for V8 ([article](https://eclipsesource.com/blogs/2016/07/25/typedarrays-sharing-memory-between-v8-and-java/)).

Regarding the copying of buffer data that TF Java performs today, I believe the rationale is that TF assumes that it owns the memory.   There's some reference counting there but I haven't looked at it recently.   Meanwhile, I am very keen on introducing [Netty-style reference counting](https://netty.io/wiki/reference-counted-objects.html) to TF Java, and intend to write a proposal to that effect.

Craigacp commented on sigs/jvm/rfcs/20190904-java-tensors.md:
```diff
@@ -0,0 +1,301 @@
+
+It has been reported that in some cases, supporting compile-time type safety can become a nightmare when 
+testing a network using different data types. Actually, this can require a lot of search & replace on each 
+experiment (e.g. `FLOAT` vs `DOUBLE`). 
```
If we expanded the set of input and output methods (i.e. so Tensor<type> had methods that accepted and returned all Java types and did the expected conversions) we could use a generic form of this, store the required enum or DataType implementation in a field and then make all tensor construction either accept that field or infer the type from other tensors. That would make it a single line change to change data types.

karllessard replied:
Yeah I'm wondering if the Tensor should accept all Java types and handle the conversion... right now, it does expose all input/output methods but fail at runtime if the type is not compatible the tensor data type, like [here](https://github.com/tensorflow/tensorflow/blob/36e4ce080e77aa1126a3d26f22a4e7bd656a7c91/tensorflow/java/src/main/java/org/tensorflow/Tensor.java#L469).

There might be other issues as well, think of "UInt32"... how do you pass such a value in a 32-bit Integer?? It sounds like we might need to some conversion in the end...

Craigacp replied:
The problem with accepting all types is when we have value types and can more easily express uint32 it'll be irritating to deal with it. But the type system will be quite different then, and we're unlikely to upgrade to that Java version for a while after release (plus it's not ready yet).

Thinking more about this, maybe there should be a static factory on Tensor that accepts a DataType and an NDArray or other mass type and performs the appropriate conversions. That way the only thing that needs to change is the data type which can be a variable stored in a single place. That's pretty close to what exists in the current version, but allowing it to make type conversions rather than just checking for type equality.

karllessard replied:
I still don’t have a solution for supporting those kind of DataTypes... again, if we let the user pick the representation of his choice for initializing the data instead of forcing him to a specific one might help a bit (e.g. a user could decide to use `int`s, `float`s or `double`s or even `BigInteger/Decimal` to feed a `TFloat` tensor, instead of forcing him to use `floats`s only). We’ll then need a converter to map the provided values to the physical buffer... 

That’s something that wasn’t foreseen though and we should work on it if we all think that’s the right approach.

Craigacp commented on sigs/jvm/rfcs/20190904-java-tensors.md:
```diff
@@ -0,0 +1,301 @@
+
+Note also that an advantage of using `NdArray` is that the type and the shape of the tensor is explicit and can be easily 
+retrieved by calling `t.dataType()` and `t.shape()` respectively, while with standard Java arrays those need
+to be inferred and discovered by the TF Java client using costly reflections and navigating through the array.
```
It's not a reflective operation that takes time, it's just the physical traversal of the Java multidimensional array, as they could be stored anywhere on the heap, and a 4d array could be made up of hundreds of thousands of arrays.

karllessard replied:
ok I'll change the text to avoid the confusion.

Craigacp commented on sigs/jvm/rfcs/20190904-java-tensors.md:
````diff
@@ -0,0 +1,301 @@
+```
+Of course, the standard Java arrays gives a better view on where is located a given row, by embedding it
+through a series of brackets. The actual proposal is focusing on simplicity and performance more that readability.
+*Please, feel free to suggest any better solution.*
````
I'm not too fond of this implicit row wise construction as I got completely lost trying to figure out where I was in the Tensor at any given call, plus it means the TensorBuilder is carrying around implicit state saying where it wants to write to next. 

At the expense of readability I think it would be preferable to use something like
```
TInt32.ofShape(3,2,2)
.set(0,0,new NDArray<>(1,2))
.set(0,1,new NDArray<>(3,4))
.set(1,0,new NDArray<>(5,6))
.set(1,1,new NDArray<>(7,8))
.set(2,0,new NDArray<>(9,10))
.set(2,1,new NDArray<>(11,12));
```
Replacing `new NDArray<>()` with whatever the appropriate constructor is. To make the varargs work we might have to permute the order of arguments (i.e. `new NDArray<>(),int..`). It could also be a TensorBuilder and require a `done()` method to construct too, I've no strong feelings either way.

karllessard replied:
hehe, yes me neither to be honest, I just drop the basic idea here to let others suggest something better :)

The example you gave is pretty much already supported by the `NdArray` API, except that the coordinates are provided after the value (because its a vararg that can conflict with an int scalar value). For example, you can set a scalar value like this:
```java
TInt32 t  = ...;
t.set(100, 0, 0).set(200, 1, 0);
```
We could add something similar but vectors, which will accept a vector, something like
```java
TInt32.ofShape(3, 2, 2)
    .set(vector(1, 2), 0, 0)
    .set(vector(3, 4), 0, 1)
    .set(vector(5, 6), 1, 0)
    ...
````
where `vector` instantiate a rank-1 array... or simply a primitive or auto-boxed array. What do you think? I understand that specifying the coordinates for each vector is more explicit but more verbose at the same time. I wouldn't mind though, personally.

Craigacp replied:
Yeah, I think that's better. I'm in favour of explicitness when constructing values.

Craigacp reviewed (commented):
I've gone through it with some initial comments. Do you have any of this refactoring implemented? I think it might be useful to look at some examples of the changes it implies in user code.

karllessard commented:
> I've gone through it with some initial comments. Do you have any of this refactoring implemented? I think it might be useful to look at some examples of the changes it implies in user code.

I partially do in the `tensor-refactoring`of that repository, which is incomplete at this moment. I'll try to do a full conversion of a whole example instead and let you know.

EronWright commented:
It strikes me that `Tensor<T>` behaves somewhat like a Scala type class, when we think of it as representing behavior (e.g. math ops) applied to `NDArray`. Might be worth adopting such idea explicitly, and/or looking at how "type classes" are typically modeled in Java.   (random thought)

karllessard commented:
> It strikes me that `Tensor<T>` behaves somewhat like a Scala type class, when we think of it as representing behavior (e.g. math ops) applied to `NDArray`. Might be worth adopting such idea explicitly, and/or looking at how "type classes" are typically modeled in Java. (random thought)

In the current proposal, a `Tensor` is simply a `NdArray` that has been allocated by TensorFlow, so they are quite interchangeable. Just to illustrate it, ˋNdArray` could have been renamed to ˋTensor` while ˋTensor` could have been renamed ˋTFTensor`.

karllessard commented:
Thinking of it, with that new proposal, if we opt for concrete classes per type (`TInt32`, ˋTString`, ...), we could probably just rename ˋNdArray` to ˋTensor` (a name that I prefer) and simply drop the actual ˋTensor<>` class.

karllessard commented:
@Craigacp : [here](https://github.com/karllessard/models/pull/1) is a little sample of what the migration would be from actual Tensor API to the new one. Note that this code does not compile yet and also assume that will find a way to avoid those `try-with-resources` everywhere, like we are discussing. 

It's not the best example though as there is not much happening here but well, at least there is one.

Craigacp commented:
WRT Scala type classes there was some discussion of typing tensors in Java Scala and kotlin at the jvmls after Erik Meijer's talk and I think the consensus was that the Java type system isn't powerful enough. That said my type theory isn't too strong and John Rose was taking about something like that on Twitter during it (https://twitter.com/JohnRose00/status/1156308406585532416?s=19).

karllessard commented:
@jxtps, do you have some opinion/suggestions on this topic?

I think it’s pretty to close to what you were discussing before in the SIG mailing list.

jxtps commented:
Hey, sorry I have been neck-deep in some other projects, will try to take a look later this week.

EronWright commented:
I think that @Craigacp has a good point that the Java type system isn't powerful enough (and in some cases may even be counter-productive).

To @karllessard's point about NDArray vs Tensor, I prefer to separate the tensor (view) vs buffer (storage) separation.  2 cents.

May I suggest you take a look at [tensorflow-scala](https://github.com/eaplatanios/tensorflow_scala), which has a fairly well-developed set of types.   At least for comparison sake.

jxtps commented:
Nice work on the RFCs @karllessard et al! I have a couple of somewhat open-ended questions / comments (sorry I'm a little all over the place): 

1. The creation of "small" tensors could use a fluency boost. `Tensors.ofDouble(new long[]{2,2})` could instead be `Tensors.ofDouble(2, 2)`, i.e. using `long... dims`, right? (This was from https://github.com/karllessard/community/blob/master/sigs/jvm/rfcs/20190606-java-tensor-io.md )
    1. Since everything needs to start from a Shape (or long[]), maybe add convenience methods on there to easily create constants, variables & tensors?
    1. Or is that too many ways to do the same thing? _There already seems to be quite a few..._ 
1. Feels like we should enumerate a couple of use-cases and make sure the APIs work smoothly for them. 
    1. Do people really construct "small-to-medium" sized tensors as constants in their code? I.e. is `TInt32.ofShape(3, 2, 2).row(...` likely to see significant use? 
    1. So far I've been working mostly with images, and I get them into tensorflow by creating a large, flat float[] that I have carefully populated, then attach a size and create a tensor. That works reasonably well, but obviously has some memcopy overhead. 
    1. Could obviously switch to using the NdArray write primitives. 
        1. I'd be a little wary of calling a `set(float value, long... indices)` type method since then you'd have to recompute the flat array index on every call. 
        1. The `Iterator<>` is probably fine, but then there's no "indexing help" (not super-needed, so this may be fine - probably better to start with a slim interface and then we can add conveniences as we discover the real need?)
1. I'm not 100% clear on what the best way to handle the various `<Type>` tags. 
    1. For starters, I still doubt the usefulness of making the `dtype` part of the type signature. 
    1. If we're going to keep it, your proposal sounds broadly feasible, though there's something that nags at me but I can't quite put my finger on it. I think I'd prefer to have a `class DataType<T_ForNdArray, T_ForTfInternals>` but I don't think that would really work since the corresponding `Tensor` type would still need to independently specify the `T_ForNdArray` and we'd want to avoid that duplication.
        1. Note that users would be able to create new DataTypes if there's a non-private `DataType.create` method, though that's probably not a big deal in reality. 
1. Avoid boxing like the plague. Use `int` instead of `Integer` wherever possible - there's boxing overhead, there are some java <-> scala interop issues (weird corner cases), and you never actually want an `Integer`, you always want an `int`. 
    1. If we're going to create all the separate primitive-type-specific classes, then there's really no need to use `Integer` et al basically anywhere.

jxtps commented:
1. With `tf.math.add(c, c)` returning a specific tensor type (`TInt32`), you'll have to overload the implementation of `tf.math.add` at least once per tensor type. 
    1. That's going to add up fast - if there are N ops and M data types, then you'll end up with O(N*M) implementations. Sure, that's in generated code, but still - you're putting a big burden on the framework code / devs for marginal user benefit, and e.g. IntelliJ auto-complete will offer dropdowns that repeat the same function M times, cluttering it up and greatly reducing its value. 

1. With `public abstract class Tensor<T> implements NdArray<T> { ... }` and `public final class TInt32 extends Tensor<Integer> implements Numeric { ... }`, how will `TInt32` offer direct `IntNdArray` functionality? 

1. What is the full mapping between `T_ForNdArray` and `dtype`?

The Tensor type-safety features you're looking to create could *maybe* work in Scala where there are typedefs, primitive specialization of generic classes, and multiple inheritance via traits. 

In Java, which is a great language in so many ways, you will be banging your head against the type system at every turn. 

Framework developer time and effort is a very limited resource. Tensor dtype safety is not *that* big of a deal in practice. Let's not create an enormous type edifice that we will be struggling to design and maintain when doing: 

```java
public class Tensor {
   final public DataType dtype;
   ... 
   public IntNdArray asInts() {...} // Throws exception if dtype isn't compatible
   ...
}
```
would get the job done in a fraction of the time/effort and be easy to use. 

The only substantive loss I see in shedding the `<T>` from `Tensor` is that some operations change the datatype, e.g `myFloats > 0.5f` produces a `boolean` tensor. Similarly `tf.image.decode_png` produces a `uint8` tensor. The first time around those can come as surprises. 

A potential workaround for that is to default to auto-casting and issuing a warning. The user can then choose to either turn off auto-cast (making it an error that throws an exception), or accept auto-cast (silencing the warning).

karllessard commented:
Thanks @jxtps, there is a lot to reply to right now but let me start with your latest post:

> With `tf.math.add(c, c)` returning a specific tensor type (`TInt32`), you'll have to overload the implementation of `tf.math.add` at least once per tensor type.

There is no need to overload the operation per tensor type. The type of the operation (and in the case of `math.add`, the type of its output) are still inferred by the input operands. So the signature remains `Add<T> add(Operand<T> x, Operand<T> y)`, the only thing that changes is that `T` instead of being `Integer` will be a `TInt32`, that you can access in eager mode via something like `Operand.data()`, i.e.
* `tf.math.add(x, y)` returns `Add<T>`, which implements `Operand<T>`.
* `tf.math.add(x, y).data()` returns a `TInt32`

> With public abstract class Tensor<T> implements NdArray<T> { ... } and public final class TInt32 extends Tensor<Integer> implements Numeric { ... }, how will TInt32 offer direct IntNdArray functionality?

That's because in this new RFC, I dropped the primitive-based `NdArray` interfaces, which ends up to be a bad decision. I restored them in a [new prototype](https://github.com/karllessard/tf-java/tree/nio-converters) I'm working on right now to prevent the problem you reported. 

In this new prototype, tensor types like `TInt32` does not extend `Tensor` anymore. Instead, `Tensor<T>` remains a reference to a native TF tensor, like now. But `T` is a tensor type instead of a java type. And it is the tensor type `T` that extends from `NdArray`, which you can access by calling `Tensor.data()`. For example,  something like
```java
Tensor<TInt32> t = TInt32.tensorOfShape(2, 2);  // TInt32 extends from IntNdArray
t.data().at(0).write(new int[] { 100, 101 })  // possible because IntNdArray accepts to write a int[]
            .at(1).write(new int[] { 102, 103 });
```

> What is the full mapping between T_ForNdArray and dtype

In this new prototype, then it is up to `dtype` to decide how it can map tensor data to `NdArray`. So in case of `TInt32`, it is pretty simple and can simply work with integers, because both are 32-bits. But, for example, `TFloat16` could still implement the `FloatNdArray` interface, while working on a `ByteNdArray`, mapping each float value to only 16-bits. I added support for this as well in the code (to be discussed, of course).

> The Tensor type-safety features you're looking to create could maybe work in Scala where there are typedefs, primitive specialization of generic classes, and multiple inheritance via traits.

If the trend is to remove the type-safety feature completely, we sure can do but we need to ask more broadly first because I think some users enjoy this aspect of TF Java. Nonetheless, the snippet you provided as a alternative is pretty close to what my new prototype is exposing, except that by keeping type-safety present, only one method is required (`data()`)  which already returns the right type of `NdArray`.

To be continued...

karllessard commented:
> The creation of "small" tensors could use a fluency boost. Tensors.ofDouble(new long[]{2,2}) could instead be Tensors.ofDouble(2, 2), i.e. using long... dims, right? 

I tried to simplify it too by moving the factory methods to the datatype class instead of `Tensors` (so you don't need to specify it). In the RFC, it is `TInt32.ofShape(2, 2)` but now I would change it to `TInt32.tensorOfShape(2, 2)`, since the return value would be a `Tensor`and not a `TInt32`.

Also, we could have some shortcuts for rank-0 and rank-2 tensors, such as `TInt32.scalar(1)` and `TInt32.vector(1, 2, 3, 4)`.

> Or is that too many ways to do the same thing? There already seems to be quite a few...

That is a very valid point, there is also the `Constant` class that builds up tensors... ideally, it should be clear for the user which one to use in which and not give him too many options...

alextp committed:
changes

alextp committed:
oops

alextp committed:
added docstring link

alextp committed:
style is a section header

alextp committed:
strings over enums

EronWright commented:
Suggestion that you close this PR if it is obsolete.

karllessard commented:
Thanks @EronWright for raising this up, while a lot of stuff is still accurate with what has been merged lately, some of it must be updated. Instead of just closing it, it could also be interesting to make changes and merge it in TF community repo so it gains some visibility to all TF developers. I can take a look at it.

Commit:
Merge pull request #171 from lzr-google/master

RFC: TPU SavedModel Export API for TF2

qlzh727 committed:
Initial commit for RFC standalone Keras repository

qlzh727 committed:
Update wording and nits.

fchollet committed:
Update 20200205-standalone-keras-repository.md

qlzh727 committed:
Merge pull request #2 from fchollet/patch-1

Proofreading, etc.

qlzh727 committed:
Update the RFC nubmer

1025KB committed:
Update 20200117-tfx-generic-trainer.md

1025KB committed:
Update 20200117-tfx-generic-trainer.md

qlzh727 committed:
Update rfcs/20200205-standalone-keras-repository.md

Co-Authored-By: Frédéric Branchaud-Charron <frederic.branchaud-charron@usherbrooke.ca>

qlzh727 committed:
Update rfcs/20200205-standalone-keras-repository.md

Co-Authored-By: Gabriel de Marmiesse <gabrieldemarmiesse@gmail.com>

qlzh727 committed:
Update tf to Keras usage with more details.

qlzh727 committed:
Change to use the list view for comparison.

Make it easy for comment to point to the exact place.

qlzh727 committed:
Update the section for PIP package dependency.

Intake the suggestion from @gabrieldemarmiesse based on the tf-addons.

qlzh727 committed:
Update the preferred PIP package version.

qlzh727 committed:
Update more details about the Github migration.

Commit:
Update status to accepted.

Commit:
Merge pull request #193 from frankchn/rfc-snapshot

RFC: tf.data Snapshot

Commit:
Clone draft template for new tf-types RFC.

Commit:
Initial commit.

Commit:
Capitalization

annarev committed:
Update rfcs/20191127-pip-structure.md

Co-Authored-By: Gabriel de Marmiesse <gabrieldemarmiesse@gmail.com>

Commit:
Update 20200211-tf-types.md

Commit:
Update 20200211-tf-types.md

Commit:
Update 20200211-tf-types.md

Commit:
Update 20200211-tf-types.md

Commit:
Update 20200211-tf-types.md

Commit:
Update 20200211-tf-types.md

Commit:
Update 20200211-tf-types.md

1025KB committed:
Update 20200117-tfx-generic-trainer.md

Commit:
Merge pull request #201 from 1025KB/patch-1

RFC: TFX Generic Trainer / Native Keras Support

Commit:
Update 20200211-tf-types.md

qlzh727 committed:
Update wording wrt to pure python project.

qlzh727 committed:
Update the sectoin for TF to keras dependency.

Commit:
Update 20200211-tf-types.md

Commit:
Update 20200211-tf-types.md

Commit:
Update 20200211-tf-types.md

genehwung committed:
Update 20200117-tfx-combining-model-validator-with-evaluator.md

bmzhao committed:
RFC: Tensorflow SavedModel C/C++ API

Commit:
Update 20200211-tf-types.md

Commit:
Update 20200211-tf-types.md

Commit:
Update PR number.

Commit:
Add `TensorLike` to core types.

Commit:
Clarify note about static checkers.

Commit:
Add note about ABC performance concerns

Commit:
Clarify semantics of `Optional`

Commit:
Add details and example for `Optional`

Commit:
Fix typo.

Commit:
Update 20200211-tf-types.md

tanzhenyu committed:
Finalize design reviews for Keras categorical inputs.

Finalize design reviews.

qlzh727 committed:
Address the comment from @apassos

1. Udpate the motivation to emphasize the modularity of TF.
2. Update the section for reverse dependency from TF to keras.

qlzh727 committed:
Update the states for python change history.

These numbers might help us decide whether the split PR execise is too
much a overhead or not.

qlzh727 committed:
Updating the section for CI and presubmit.

qlzh727 committed:
Update style and fix typo.

qlzh727 committed:
Adding a section for Alternative considered.

angerson committed:
Update @angersson to @angerson (#210)

Since my username has changed, this gets a fix!

qlzh727 committed:
Fix typo for LasyLoader

VoVAllen committed:
Update 20191016-dlpack-support.md

EvenOldridge committed:
Merge pull request #2 from VoVAllen/patch-1

Update 20191016-dlpack-support.md with VoVAllen's changes

genehwung committed:
Update 20200117-tfx-combining-model-validator-with-evaluator.md

Commit:
Merge pull request #200 from genehwung/master

RFC: Combining TFX ModelValidator with TFX Evaluator

tanzhenyu committed:
Update 20191212-keras-categorical-inputs.md

Finalize design and add review meetings.

EvenOldridge committed:
Update 20191016-dlpack-support.md

Removed the questions.

VoVAllen committed:
Update 20191016-dlpack-support.md

Commit:
Add discussion topic

qlzh727 committed:
Update the RFC wrt to the design meeting outcome.

qlzh727 committed:
Update the RFC status to 'Accepted'

EvenOldridge committed:
Merge pull request #3 from VoVAllen/patch-2

Update 20191016-dlpack-support.md

Commit:
Clarify intent for type checking mechanics

Commit:
Merge pull request #202 from qlzh727/master

RFC: Standalone Keras Repository

EvenOldridge committed:
Update 20191016-dlpack-support.md

Removed request for contributions, added thinc.ai framework interoperability.

VoVAllen committed:
Update 20191016-dlpack-support.md

EvenOldridge committed:
Merge pull request #4 from VoVAllen/patch-3

Update 20191016-dlpack-support.md

Commit:
Add @ematejska to code owners for sig/swift (#212)

rachellj218 committed:
RFC: TensorFlow Official Model Garden Redesign (#130)

This should have been merged some time ago. The design is undergoing iteration from this original RFC. Thank you everyone for your conversation and contribution.


* Create 20190802-model-garden-redesign.md

* Update 20190802-model-garden-redesign.md

* Update 20190802-model-garden-redesign.md

* Update 20190802-model-garden-redesign.md

* Update 20190802-model-garden-redesign.md

* Update 20190802-model-garden-redesign.md

* Status -> Accepted

Co-authored-by: Edd Wilder-James <ewj@google.com>

Commit:
Initial RFC for single-client parameter server training.

Commit:
Remove rebuild_arbitrary_future.png

Commit:
Use a smaller png.

Commit:
Fix small typos.

rchao committed:
Fix extra period. 

Fix extra period.

Commit:
Merge pull request #1 from rchao/patch-1

Fix extra period.

Commit:
Add section on protocol for custom Tensors

Commit:
Add open question about legacy API

Commit:
Add discussion topic for tf.TypeSpec

Commit:
Add one more paragraph for single client

alextp committed:
changes

alextp committed:
s/tf-api/owners/ tensorflow/api-owners'

martinwicke committed:
final touches

alextp committed:
..

alextp committed:
.

alextp committed:
Merge pull request #185 from alextp/api-reviews

Criteria for doing API reviews and design reviews in TF

alextp committed:
experimental API policy

alextp committed:
clarifying known deficiency

Commit:
Detail the relationship with tf.TypeSpec

Commit:
Update based PR comments.

Commit:
Formatting

Commit:
Draft guidance for adding new types

Commit:
Minor edit

alextp committed:
Update governance/api-reviews.md

Co-Authored-By: Martin Wicke <577277+martinwicke@users.noreply.github.com>

alextp committed:
Update governance/api-reviews.md

Co-Authored-By: Martin Wicke <577277+martinwicke@users.noreply.github.com>

alextp committed:
clarifications

Commit:
Use a proper example for type annotations

Commit:
Update 20200211-tf-types.md

Commit:
Fix typo.

Commit:
Clarify that example is non-prescriptive

Commit:
Update status to proposed

Commit:
Update time stamp

Commit:
Fix typo

Commit:
Slight update to example

Commit:
Add link to an example

Commit:
Add design review notes

Commit:
Fix indentation in notes

Commit:
Minor updates following design review

Commit:
Update status

Commit:
Merge pull request #208 from mdanatg/master

RFC: TensorFlow Canonical Type System

jsimsa committed:
Optional arguments with defaults.

jsimsa committed:
Fix typo.

seanpmorgan committed:
* Updated procedure (#203)

Commit:
Update 20200306-single-client-parameter-server.md

alextp committed:
clarify policy for stuff which keeps changing

alextp committed:
reference experimental section from other places

alextp committed:
clarify addons

liutongxuan committed:
RFC: FuseRecv

Commit:
Update 20200306-single-client-parameter-server.md

Commit:
Change the status to accepted

Commit:
Merge pull request #214 from yuefengz/master

Single-Client Parameter Server Training

Commit:
Merge pull request #209 from tanzhenyu/master

RFC: Keras categorical inputs

jbedorf committed:
Fix irregular writing and expression issues.

liutongxuan committed:
update reviewers and submit time

liutongxuan committed:
update sponsor

LexusH committed:
Update faq.md (#205)

Fixed 6 broken links mentioned in Issue #191

ashutosh1919 committed:
Fixed broken link for Bug & Performance issues (#216)

Commit:
Merge pull request #182 from annarev/pip_structure_rfc

RFC: Improved pip package structure

Commit:
Mark RFC as implemented

Commit:
Merge pull request #228 from tensorflow/ematejska-patch-2

Mark RFC as implemented

Commit:
Merge pull request #169 from underscoreanuj/master

Update 20180918-functions-not-sessions-20.md

Commit:
Updated this to accepted.

Commit:
Merge pull request #180 from EvenOldridge/EvenOldridge-dlpack-support

RFC: DLpack support for interoperability with other GPU frameworks

liutongxuan committed:
fix typo

jsimsa committed:
Update CHARTER.md (#229)

mmilanifard committed:
Updates after design review. Design accepted.
TF Lattice is launched at https://www.tensorflow.org/lattice.

Commit:
Merge pull request #186 from mmilanifard/master

RFC: TensorFlow Lattice 2.0

Commit:
Create SIG Graphics charter

Commit:
Update CHARTER.md

typos

Commit:
Merge pull request #231 from tensorflow/theadactyl-patch-1

Create SIG Graphics charter

Commit:
Update CHARTER.md

Added membership details, fixed typos

terrytangyuan committed:
Fix markdown formatting in SIG Graphics charter

Commit:
Merge pull request #233 from terrytangyuan/patch-1

Fix markdown formatting in SIG Graphics charter

Commit:
Update to set the status to proposed.

Commit:
Updating the Proposed status to Accepted for these merged RFCs. (#235)

Commit:
Merge pull request #194 from arnoegw/patch-1

Fix typo in 20190430-tokenization-conventions.md.

JaccovG committed:
Update dead link in charter.md (#189)

the micro folder was moved out of the experimental folder. this was not yet updated in the link.

samikama committed:
Introducing Transactions extension to Modular Filesystems RFC

Commit:
Merge branch 'master' into theadactyl-patch-2

1025KB committed:
Add files via upload

1025KB committed:
Merge pull request #1 from 1025KB/1025KB-patch-2-1

1025 kb patch 2 1

1025KB committed:
Delete cloud.png

1025KB committed:
Delete parallel_tuning.png

1025KB committed:
Delete workflow.png

1025KB committed:
Create 123

1025KB committed:
Add files via upload

1025KB committed:
Delete 123

1025KB committed:
Update 20200420-tfx-tuner-component.md

Commit:
Mark as Accepted.

Commit:
Merge pull request #207 from bmzhao/tf-c-saved-model

RFC: Tensorflow SavedModel C/C++ API

aaudiber committed:
Add C++ style guide.

This guide is a starting point for documenting Tensorflow C++ style,
especially where it differs from general Google C++ style.

Commit:
Merge pull request #232 from tensorflow/theadactyl-patch-2

Update CHARTER.md

liutongxuan committed:
Update user API and op definition based on comments.

Commit:
Merge pull request #236 from 1025KB/master

RFC: TFX Tuner Component

Commit:
Merge pull request #224 from liutongxuan/master

RFC: FuseRecv

penpornk committed:
Add CSRSparseMatrix RFC.

penpornk committed:
Adjust image sizes.

penpornk committed:
Fix an image's path.

penpornk committed:
Add a discussion point about CompositeTensor.

penpornk committed:
Update the RFC #.

reedwm committed:
Create RFC for TensorFloat-32 support in TensorFlow

reedwm committed:
Update RFC link to point to PR

Also fix date format

reedwm committed:
Address review comments.

reedwm committed:
Address review comments

reedwm committed:
Address comments

samikama committed:
Added updates from discussions

seanpmorgan committed:
* gelu migration RFC

seanpmorgan committed:
typos

seanpmorgan committed:
Spacing

seanpmorgan committed:
* Update transition

penpornk committed:
Add information about documentation and tutorial.

reedwm committed:
Add some more info on remote devices

penpornk committed:
Clarify why COO is not efficient with common TF sparsities

seanpmorgan committed:
Add Keras activations

seanpmorgan committed:
Add RFC number

annarev committed:
Removing device part of the proposal

I removed device section of this doc. We never made a final decision on the device solution and then I moved to work on something else. If we come up with a final decision for device, we can create another RFC (I removed my name from authors as well since my only contribution was device API. ).

The rest of the doc can be merged.

annarev committed:
Changed "Updated" date

Commit:
Marked as accepted.

Commit:
Merge pull request #247 from reedwm/tf32

RFC: TensorFloat-32 support in TensorFlow

Commit:
Moving to Accepted

Commit:
Merge pull request #133 from sjamesr/add_kernel_rfc

RFC: Kernel and Op Implementation and Registration API

ruoyu90 committed:
RFC for advanced TFX DSL semantics

samikama committed:
Remove unnecessary token arguments from stateful tokens example

ruoyu90 committed:
Fixed format

ruoyu90 committed:
Fixes table format.

1025KB committed:
Update 20200420-tfx-tuner-component.md

Commit:
Merge pull request #256 from 1025KB/patch-2

Update 20200420-tfx-tuner-component.md

saberkun committed:
Create 20200616-keras-multihead-attention.md

The feedback phase will be open for two weeks until Wednesday July 02, 2020.

# RFC: Multihead Attention and EinsumDense on Keras

| Status        | (Proposed / Accepted / Implemented / Obsolete)          |
| :------------ | :------------------------------------------------------ |
| **RFC #**     | [NNN](https://github.com/tensorflow/community/pull/NNN) |
:               : (update when you have community PR #)                   :
| **Author(s)** | Hongkun Yu (hongkuny@google.com)                        |
| **Sponsor**   | Francois Chollet (fchollet@google.com)                  |
| **Updated**   | 2020-06-16                                              |

## Objective

Introduce the MultiHeadAttention layer and EinsumDense layer to tf.keras.

saberkun committed:
Update 20200616-keras-multihead-attention.md

saberkun committed:
Update 20200616-keras-multihead-attention.md

    `tf.keras.layers.experimental.MultiHeadAttention` and
    `tf.keras.layers.experimental.EinsumDense`

saberkun committed:
Update 20200616-keras-multihead-attention.md

Add mark to authors.
Add plan for addons migration.

saberkun committed:
Update 20200616-keras-multihead-attention.md

saberkun committed:
Use the mask_tensor inside the example code.

Commit:
This has been accepted.

penpornk committed:
Update the RFC based on comments.

ruoyu90 committed:
Updating future works to emphasize `TriggerPolicy`

Commit:
Marking as Accepted.

Commit:
Merge pull request #253 from ruoyu90/async

RFC: TFX Advanced DSL semantics

penpornk committed:
Add references for deep learning sparsities.

Commit:
Merge pull request #246 from penpornk/master

RFC: CSRSparseMatrix

saberkun committed:
Update 20200616-keras-multihead-attention.md

samikama committed:
Adding modifications for merged proposal

Commit:
Update 20200505-transactional-fs.md

Commit:
Update 20200616-keras-multihead-attention.md

samikama committed:
Added transactional checkpointing example and possible implementation changes

tanzhenyu committed:
Update Keras categorical input design, mark it as complete.

More importantly, add an official guide on how TF1 Estimator users with feature columns can convert their code into TF2 Keras users with KPL.

tanzhenyu committed:
Address comments.

tanzhenyu committed:
Fix some punctuation.

Commit:
Merge pull request #267 from tanzhenyu/master

Update Keras categorical input design, mark it as complete.

Commit:
Mark as Accepted

Commit:
Merge pull request #245 from samikama/sami_transactional_fs

RFC: Introducing Transactions extension to Modular Filesystems

seanpmorgan committed:
* Marking RFC as completed

Commit:
Merge pull request #252 from seanpmorgan/gelu-migration

RFC: Migrate gelu activation from Addons to Core

saberkun committed:
Update two proposed changes

Update two proposed changes to the existing Attention layer

saberkun committed:
Update to accepted

Commit:
Merge pull request #260 from saberkun/patch-1

RFC: Multihead Attention and EinsumDense on Keras

edloper committed:
Merge pull request #1 from tensorflow/master

Merge changes from base

edloper committed:
Adding RFC: TF Extension Types

edloper committed:
Updated markdown to enable python code formatting.

edloper committed:
markup fixes

edloper committed:
markup fixes

guptapriya committed:
Policy to require public types for API

We want to require all new public APIs to have publicly documented arguments and return values.

guptapriya committed:
Updated wording

Commit:
Update 20200721-extension-types.md

edloper committed:
minor updates; marked RFC as accepted

edloper committed:
Merge branch 'master' of github.com:edloper/community

Commit:
Merge pull request #269 from edloper/master

RFC: TensorFlow Extension Types
